### PR TITLE
Callable Layers and NeighborList layers

### PR DESCRIPTION
**Pull request by rbharath**
This PR proposes a modification to the `TensorGraph` API that makes layers callable:

- Adds `TensorWrapper`, a subclass of `Layer` that wraps arbitrary tensorflow Tensors in dummy layers that have only `out_tensor` set.
  - `wrap = TensorWrapper(some_tensor)`
- Adds a methods `__call__(*in_layers)`. Underneath the hood, this method sets `self.in_layers` and then invokes the `_create_tensor` method to create tensors. If inputs are `tf.Tensor` objects, wraps them with `TensorWrapper` first.
  - Complex layers are hard to unit-test in the current framework. Making layers callable allows for easy testing since `Layer` objects can be invoked on `tf.Tensor` objects and tested just like standard tensorflow functions using `TensorflowTestCase`.
  - `Layer` objects can now be invoked easily in standalone Tensorflow code. Allows users to make use of DeepChem layers without having to buy-in to the full DeepChem workflow.
- PR uses this API (and standard `TensorGraph`) to implement a simple `Vina` score function.
- Adds a minor change to `Databag` allows `datasets` to be passed in the constructor. Cuts boilerplate:
  - `databag = Databag({prot_coords: X_prot, ligand_coords: X_ligand, labels: y})`
- Adds `axis` arguments to layers `ReduceSum`, `ReduceMean`, `ReduceSquareDifference` to allow for greater control.

@lilleswing and @peastman: Please take a look and let me know your thoughts here. Hope to convince you that callable layers allow for elegant constructions :)

### CODE REVIEW

**Comment by peastman** [comment]
I haven't had time to look at it yet.  I'll try to do that later today.

**Comment by peastman** [comment]
As we discussed offline, I really don't like this API.  Treating layers as functions is just confusing.  What does it mean to "call" a layer?  What are the arguments?  What is the return value?  And you would never guess that calling it as a function actually modifies the layer object itself.  It's totally unintuitive.

A much better API would be to have an explicit method that returns a tensor without modifying the layer object.  A simple way of doing that would be to just add an optional `in_layers` argument to `create_tensor()`.  So to take an example from one of the test cases, instead of

```Python
hbond = VinaHydrogenBond()(X_tensor)
```

you would write

```Python
hbond = VinaHydrogenBond().create_tensor(in_layers=[X_tensor])
```

It's a little more wordy, but much clearer, and it doesn't have any surprising side effects.

**Comment by peastman** [comment]
I'm a little worried about the number of super-specialized layers piling up in layers.py that are entirely specific to one particular model.  Tensorgraph is supposed to be a generic graph API, so having classes like `VinaGaussianSecond` in it feels wrong.  If a layer is really only intended to be used in one particular model and nowhere else, it should be defined inside that model.

We might also think about creating a `CustomLayer` class to simplify creating really simple layers like this.  For example, it might use a lambda to define its output:

```Python
layer = CustomLayer(output=lambda x: tf.exp(-((x[0]-3)/2)**2))
```

**Comment by rbharath** [comment]
I agree that the side-effect is surprising and worth fixing. Adding an optional `in_layers` argument to `_create_tensor` (and making the method public as `create_tensor`) seems like a good fix to me. I'll go ahead and make these changes in the PR.

With documentation though, having callable layers serve as a syntactic shorthand for `.create_tensor(in_layers)` is valuable. Brevity can be very useful when trying to fit complicated networks into the ML designer's head.

+1 for `CustomLayer`. I also agree that the build-up of tiny classes isn't that helpful. Should be straightforward to refactor those into a single `Vina` layer since I don't expect them to be used elsewhere.

**Comment by peastman** [comment]
> Should be straightforward to refactor those into a single Vina layer since I don't expect them to be used elsewhere.

Whether it's one layer or many isn't important.  The main thing is to define them in vina_model.py instead of layers.py.

File: deepchem/models/tensorgraph/layers.py [hunk 1]
```diff
@@ -69,6 +69,26 @@ def shared(self, in_layers):
...
+          raise ValueError("Layer must be invoked on layers or tensors")
+      self.in_layers = layers
+    return self._create_tensor()
+
+
+class TensorWrapper(Layer):
+  """Used to wrap a tensorflow tensor."""
+
+  def __init__(self, out_tensor):
+    self.out_tensor = out_tensor
```
Comment (peastman) [inline deepchem/models/tensorgraph/layers.py]: Doesn't TensorWrapper need to implement `create_tensor()`?  Otherwise you'll get an exception when you try to use it.

Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Hmm, good point. Will we ever want to use `TensorWrapper` in a `TensorGraph` or only when layers are invoked in pure tensorflow?

I'll go ahead and add in a `create_tensor` since it will be easy though.

**Comment by rbharath** [comment]
Updated this PR to:

- Add optional `in_layers` argument to `create_tensor` (which is now a public method)
- Make `__call__` syntactic sugar for `create_tensor`
- Add unit tests for most implemented `TensorGraph` layers. These simply test that the output shape of each layer is correct.

Will merge in once unit tests pass.

**Comment by rbharath** [comment]
Forgot to mention:

- This PR simplifies names of certain layers: `BatchNormLayer` -> `BatchNorm`, etc. (removes the redundant `Layer`).

**Comment by peastman** [comment]
> I agree that the side-effect is surprising and worth fixing.

It looks like you never fixed this.  If you call `create_tensor()`, passing in a list of layers, it still modifies `self.out_tensor`.

**Comment by rbharath** [comment]
Good point. I fixed the self.in_layers modification, but forgot to handle
self.out_tensor. I'll fix this in a follow-up PR.


On Apr 26, 2017 11:08 AM, "peastman" <notifications@github.com> wrote:

I agree that the side-effect is surprising and worth fixing.

It looks like you never fixed this. If you call create_tensor(), passing in
a list of layers, it still modifies self.out_tensor.

—
You are receiving this because you modified the open/close state.

Reply to this email directly, view it on GitHub
<https://github.com/deepchem/deepchem/pull/523#issuecomment-297494860>, or mute
the thread
